### DISCUSSION TITLE
Add snmp inventory Switch Huawei

### DISCUSSION THREAD

**Discussion by xfilho** [glpi-project/glpi-agent#327] [Q&A]
### Bug reporting acknowledgment

Yes, I read it

### Professional support

None

### Describe the bug

I'm trying to add the huawei switch model S5720-52X-PWR-SI-AC in glpi using netdiscovery/netinventory, but I'm not succeeding. The agent probably does not have the device's SNMP identifier in its internal database. I added to the sysobject.ids file "2011.2.23.335 Huawei NETWORKING S5720-52X-PWR-SI-AC" as per documentation, but that didn't work either.

Error:
```
[debug] [thread 1] #1, full match for sysobjectID .1.3.6.1.4.1.2011.2.23.335 in database
[warning] Aborting netdiscovery task as it reached expiration time
[error] 1 devices scan result missed
[error] job 1 aborted
[debug] All netdiscovery threads terminated
```

### To reproduce

1. On the glpi server, where the agent is installed, I executed the command: `glpi-netdiscovery --credentials version:2c,community:********* --debug --host IP-SWITCH`

### Expected behavior

I can't inventory the huawei switch S5720-52X-PWR-SI-AC

### Operating system

Linux

### GLPI Agent version

1.4

### GLPI version

10.0.x (See additional context below)

**Comment by g-bougard** [comment]
Please people, stop open a bug issue if you don't provide evidence of a bug. First open a Q&A discussion.

@xfilho please provide the full output in debug2 with:
```
glpi-netdiscovery --credentials version:2c,community:********* --debug --debug --host IP-SWITCH
```
When you see a `it reached expiration time` warning, you should also say in how many time that expiration was reached... but probably the full requested output will tell it.

Also, you should firstly not provide the output after a `sysobject.ids` update or you are no more complaining about the same product. If you see a `partial match` message (and you should) when `sysobject.ids` is not modified, this is also normal. The only change you're doing when you modify sysobject.ids for such device is to set the MODEL if it's not set.

**Reply by xfilho** [reply]
Sorry, this is the first time i open a problem. 

Output:

```
[debug] Logger backend Stderr initialized
[debug] GLPI Agent (1.4-1)
[debug] initializing job 1
[debug] initializing block IP-SWITCH-IP-SWITCH
[debug] Current run expiration timeout: soon
[debug] creating 1 worker threads
[debug] [thread 1] creation
[debug] [thread 1] #1, scanning IP-SWITCH
[debug] [thread 1] #1, full match for sysobjectID .1.3.6.1.4.1.2011.2.23.335 in database
[debug2] [thread 1] #1, /etc/glpi-agent/toolbox.yaml configuration not found
[warning] Aborting netdiscovery task as it reached expiration time
[error] 1 devices scan result missed
[error] job 1 aborted
[debug] All netdiscovery threads terminated
```

**Reply by g-bougard** [reply]
Hi @xfilho 
and how many time does it take ?
What is your OS name and its version ?
It seems it detects snmp support as you have a match on sysobjectID but it doesn't report snmp success before reaching the expiration. You may try to use `--timeout` option to give more time for snmp request answers to reach the agent.

**Reply by xfilho** [reply]
Hi, @g-bougard 

**how many time does it take ?**   30s
**What is your OS name and its version ?** CentOS Linux release 7.9.2009 (Core)
**Try to use --timeout:** 

**Command:**
```
glpi-netdiscovery --credentials version:2c,community:******** --timeout 60 --debug --debug --host IP-SWITCH
```
**Output:**
```
[debug] Logger backend Stderr initialized
[debug] GLPI Agent (1.4-1)
[debug] initializing job 1
[debug] initializing block IP-SWITCH-IP-SWITCH
[debug] Current run expiration timeout: soon
[debug] creating 1 worker threads
[debug] [thread 1] creation
[debug] [thread 1] #1, scanning IP-SWITCH
[debug] [thread 1] #1, full match for sysobjectID .1.3.6.1.4.1.2011.2.23.335 in database
[debug2] [thread 1] #1, /etc/glpi-agent/toolbox.yaml configuration not found
[warning] Aborting netdiscovery task as it reached expiration time
[error] 1 devices scan result missed
[error] job 1 aborted
[debug] All netdiscovery threads terminated
```

**Reply by g-bougard** [reply]
There's probably an issue with that old OS version. There's maybe a multi-threading support issue. This happened in the past on CentOS.
Can you try with a more recent OS or even another linux distro ?

**Reply by xfilho** [reply]
I tested using the agent on windows 10

```
> glpi-netdiscovery --credentials version:2c,community:******** --timeout 60 --debug --debug --host IP-SWITCH
[debug] Logger backend Stderr initialized
[debug] GLPI Agent (1.4)
[debug] initializing job 1
[debug] initializing block IP-SWITCH-IP-SWITCH
[debug] Current run expiration timeout: soon
[debug] creating 1 worker threads
[debug] [thread 1] creation
[debug] [thread 1] #1, scanning IP-SWITCH
[debug] [thread 1] #1, partial match for sysobjectID .1.3.6.1.4.1.2011.2.23.335 in database: unknown device ID
[debug2] [thread 1] #1, C:/Program Files/GLPI-Agent/etc/toolbox.yaml configuration not found
[warning] Aborting netdiscovery task as it reached expiration time
[error] 1 devices scan result missed
[error] job 1 aborted
[debug] All netdiscovery threads terminated
```

**Reply by g-bougard** [reply]
Humpf, maybe this device has a issue with snmp support, starting to communicate and finally stops to answer because of some restriction or firewall rule. This is the first time I see that.

**Reply by xfilho** [reply]
I'm testing with the snmpwalk from another computer with Debian and it works: 

**Command:**
snmpwalk -v2c -c COMMUNITY IP-SWITCH

**Result:**
```
iso.3.6.1.2.1.1.1.0 = STRING: "S5720-52X-PWR-SI-AC
Huawei Versatile Routing Platform Software
 VRP (R) software,Version 5.160 (S5720 V200R009C00SPC500)
 Copyright (C) 2007 Huawei Technologies Co., Ltd."
iso.3.6.1.2.1.1.2.0 = OID: iso.3.6.1.4.1.2011.2.23.335
iso.3.6.1.2.1.1.3.0 = Timeticks: (1268052537) 146 days, 18:22:05.37
iso.3.6.1.2.1.1.4.0 = STRING: "R&D Beijing, Huawei Technologies co.,Ltd."
iso.3.6.1.2.1.1.5.0 = STRING: "HUPI-SW-1604"
iso.3.6.1.2.1.1.6.0 = STRING: "POSTO 04"
iso.3.6.1.2.1.1.7.0 = INTEGER: 78
iso.3.6.1.2.1.2.1.0 = INTEGER: 58
iso.3.6.1.2.1.2.2.1.1.1 = INTEGER: 1
iso.3.6.1.2.1.2.2.1.1.2 = INTEGER: 2
iso.3.6.1.2.1.2.2.1.1.3 = INTEGER: 3
iso.3.6.1.2.1.2.2.1.1.4 = INTEGER: 4
iso.3.6.1.2.1.2.2.1.1.5 = INTEGER: 5
iso.3.6.1.2.1.2.2.1.1.6 = INTEGER: 6
iso.3.6.1.2.1.2.2.1.1.7 = INTEGER: 7
iso.3.6.1.2.1.2.2.1.1.8 = INTEGER: 8
iso.3.6.1.2.1.2.2.1.1.9 = INTEGER: 9
iso.3.6.1.2.1.2.2.1.1.10 = INTEGER: 10
iso.3.6.1.2.1.2.2.1.1.11 = INTEGER: 11
iso.3.6.1.2.1.2.2.1.1.12 = INTEGER: 12
iso.3.6.1.2.1.2.2.1.1.13 = INTEGER: 13
iso.3.6.1.2.1.2.2.1.1.14 = INTEGER: 14

```

**Reply by g-bougard** [reply]
Then can you provide the [full snmpwalk output as explain in doc](https://glpi-agent.readthedocs.io/en/latest/bug-reporting.html#snmpwalk-output) ? Maybe this could help to identify a bug that blocks the netinventory but doesn't trigger an error.

**Reply by xfilho** [reply]
https://1drv.ms/u/s!ApQNAwhNLrRclQ6khRVa_Uqcz5N4?e=0zxHZI

**Reply by g-bougard** [reply]
Thank you @xfilho 
Anyway, I can't reproduce any trouble with the walk. I can use it with `glpi-netdiscovery` & `glpi-netinventory` "--file" option without problem to generate expected XMLs.
```
[g-bougard@rog glpi-agent]$ bin/glpi-netdiscovery --file walks/huawei/S5720-52X-PWR-SI-AC.walk --debug --debug
[debug] Logger backend Stderr initialized
[debug] GLPI Agent (1.4-1)
[debug] initializing job 1
[debug] initializing block 1.1.1.1-1.1.1.1
[debug] Current run expiration timeout: soon
[debug] creating 1 worker threads
[debug] [thread 1] creation
[debug] [thread 1] #1, scanning 1.1.1.1
[debug] [thread 1] #1, partial match for sysobjectID .1.3.6.1.4.1.2011.2.23.335 in database: unknown device ID
[debug2] [thread 1] #1, /etc/glpi-agent/toolbox.yaml configuration not found
[debug] [thread 1] #1, - scanning 1.1.1.1 with SNMP, credentials 1: success
<?xml version="1.0" encoding="UTF-8" ?>
<REQUEST>
  <CONTENT>
    <DEVICE>
      <AUTHPORT />  <AUTHPROTOCOL />  <AUTHSNMP>1</AUTHSNMP>
      <CONTACT>R&amp;D Beijing, Huawei Technologies co.,Ltd.</CONTACT>
      <DESCRIPTION>S5720-52X-PWR-SI-AC 
Huawei Versatile Routing Platform Software 
 VRP (R) software,Version 5.160 (S5720 V200R009C00SPC500) 
 Copyright (C) 2007 Huawei Technologies Co., Ltd.</DESCRIPTION>
      <FIRMWARE>Version 5.160 V200R009C00SPC500</FIRMWARE>
      <IP>1.1.1.1</IP>
      <IPS>
        <IP>10.152.232.214</IP>
        <IP>127.0.0.1</IP>
        <IP>192.168.0.1</IP>
      </IPS>
      <LOCATION>POSTO 04</LOCATION>
      <MAC>48:fd:8e:0f:65:60</MAC>
      <MANUFACTURER>Huawei</MANUFACTURER>
      <SERIAL>2102350DLXDMGA001206</SERIAL>
      <SNMPHOSTNAME>HUPI-SW-1604</SNMPHOSTNAME>
      <TYPE>NETWORKING</TYPE>
      <UPTIME>149 days, 18:07:29.02</UPTIME>
    </DEVICE>
    <MODULEVERSION>5.1</MODULEVERSION>
    <PROCESSNUMBER>1</PROCESSNUMBER>
  </CONTENT>
  <DEVICEID>foo</DEVICEID>
  <QUERY>NETDISCOVERY</QUERY>
</REQUEST>
[debug2] [thread 1] processed 1 scans
[debug] [thread 1] termination
[debug] All netdiscovery threads terminated
[g-bougard@rog glpi-agent]$
```
Can you tell us which OS you're using and in which version ?

**Reply by tech62** [reply]
Hello !
I've the same issue, the inventory is running normally but when the 1 hour is reached :
```
[warning] Aborting netinventory job as it reached expiration time
[error] 1 devices inventory are missing
[error] job 1 aborted
[debug] All netinventory threads terminated
```

Is it possible to increase the 1 hour limit of glpi-netinventory.bat (with a --time-inventory-limit switch per example)?
Edit :
To make my inventory possible, i've modified the $global_timeout in the file "NetInventory.pm" located in perl\agent\GLPI\Agent\Task - line 181
`$global_timeout = 18000 if $global_timeout < 18000;`

**Reply by g-bougard** [reply]
Hi @tech62 
as you seem to point out, the problem is the set expiration, not the minimal one. I guess this would have worked if you just have changed the previous line to:
```
my $target_expiration = 18000;
```
Anyway, I'm updating the code to make the target expiration a function of `backend-collect-timeout` configuration, actually only used for inventory task.

**Reply by tech62** [reply]
Hi @g-bougard !
Thanks for the fix !
The informations are not correct for some ports, how can i use specific mibs with glpi-netinventory from windows ?

**Reply by g-bougard** [reply]
@tech62 can you explain what is not correct for these ports ?

For you to know, I just pushed 64e405c to make expiration a function of `backend-collect-timeout`. The commit also add the option to `glpi-netdiscovery` & `glpi-netinventory` scripts. To not change the default which works for most people, for netinventory task, I just multiplied the value by 5 (180 by default means 900s, so 15 minutes). So in your case, you'll have to set it to `3600` to have 18000 as resulting expiration.

Next nightly build will include the update.

**Reply by tech62** [reply]
Nice!, 
Wrong vlans returned per example

**Reply by g-bougard** [reply]
To handle such problem, you'll have to provide a [full snmpwalk output with right format](https://glpi-agent.readthedocs.io/en/latest/bug-reporting.html#snmpwalk-output) and list everything wrong in your point of view: gotten value vs expected one.
